src/server/ua_server_ns0_gds.c

```diff
@@ -0,0 +1,1124 @@
+        UA_Variant_setArray(&output[0], NULL, rejectedListSize, &UA_TYPES[UA_TYPES_BYTESTRING]);
+        return UA_STATUSCODE_GOOD;
+    }
+
+    UA_ByteString *rejectedList = (UA_ByteString*)UA_Array_new(rejectedListSize, &UA_TYPES[UA_TYPES_BYTESTRING]);
+    if(rejectedList == NULL) {
+        UA_Array_delete(rejectedListSecureChannel, rejectedListSecureChannelSize, &UA_TYPES[UA_TYPES_BYTESTRING]);
+        UA_Array_delete(rejectedListSession, rejectedListSessionSize, &UA_TYPES[UA_TYPES_BYTESTRING]);
+        return UA_STATUSCODE_BADOUTOFMEMORY;
+    }
+
+    memcpy(rejectedList, rejectedListSecureChannel, rejectedListSecureChannelSize * sizeof(UA_ByteString));
+    memcpy(rejectedList + rejectedListSecureChannelSize, rejectedListSession, rejectedListSessionSize * sizeof(UA_ByteString));
+
+    UA_Variant_setArrayCopy(&output[0], rejectedList,
+                        rejectedListSize, &UA_TYPES[UA_TYPES_BYTESTRING]);
+
+    UA_Array_delete(rejectedListSecureChannel, rejectedListSecureChannelSize, &UA_TYPES[UA_TYPES_BYTESTRING]);
+    UA_Array_delete(rejectedListSession, rejectedListSessionSize, &UA_TYPES[UA_TYPES_BYTESTRING]);
+    UA_free(rejectedList);
+
+    return retval;
+}
+
+static UA_StatusCode
+addCertificate(UA_Server *server,
+               const UA_NodeId *sessionId, void *sessionHandle,
+               const UA_NodeId *methodId, void *methodContext,
+               const UA_NodeId *objectId, void *objectContext,
+               size_t inputSize, const UA_Variant *input,
+               size_t outputSize, UA_Variant *output) {
+
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+
+    /*check for input types*/
+    if(!UA_Variant_hasScalarType(&input[0], &UA_TYPES[UA_TYPES_BYTESTRING]) || /*Certificate*/
+       !UA_Variant_hasScalarType(&input[1], &UA_TYPES[UA_TYPES_BOOLEAN])) /*IsTrustedCertificate*/
+        return UA_STATUSCODE_BADTYPEMISMATCH;
+
+    UA_ByteString certificate = *(UA_ByteString *)input[0].data;
+    UA_Boolean isTrustedCertificate = *(UA_Boolean *)input[1].data;
+
+    if(!isTrustedCertificate || certificate.length == 0)
+        return UA_STATUSCODE_BADCERTIFICATEINVALID;
+
+    /* TODO: check if TrustList Object is already open */
+    /* TODO: check if TrustList Object is read only */
+
+    UA_TrustListDataType trustList;
+    memset(&trustList, 0, sizeof(UA_TrustListDataType));
+    UA_ByteString certificates[1];
+    certificates[0] = certificate;
+
+    trustList.specifiedLists = UA_TRUSTLISTMASKS_TRUSTEDCERTIFICATES;
+    trustList.trustedCertificates = certificates;
+    trustList.trustedCertificatesSize = 1;
+
+    UA_CertificateGroup *certGroup = getCertGroup(server, objectId);
+    if(certGroup == NULL) {
+        return UA_STATUSCODE_BADINVALIDARGUMENT;
+    }
+
+    UA_CertificateGroup *transactionCertGroup =
+        UA_Transaction_getCertificateGroup(server->transaction, certGroup);
+    if(transactionCertGroup == NULL) {
+        return UA_STATUSCODE_BADINTERNALERROR;
+    }
```

Comment:
As far as I understand it, both the "AddCertificate" and "RemoveCertificate" Methods are supposed to work atomically, bypassing the transaction mechanism completely. As they do not return an "applyChanges" boolean, a user would not be able to derive if the Method started a transaction and hence requires a subsequent call to "applyChanges" later - or not.
What is more, the specification states that these functions shall abort with Bad_TransactionPending when a transaction is currently active on the specific CertificateGroup, see
https://reference.opcfoundation.org/GDS/v105/docs/7.8.2.4

That said, I think the code should rather look like this:
`    /* abort when a transaction is active on this CertGroup */
    if(UA_GDSTransaction_containsCertificateGroup(server->transaction, certGroup->certificateGroupId)) {
        return UA_STATUSCODE_BADTRANSACTIONPENDING;
    }
`